### PR TITLE
Implement Auto-Retry Mechanism for Python Code Generation

### PR DESCRIPTION
**Pull request by gengliangwang**
Introduce auto-retry framework for python code generation. The framework can be used on:
* df.ai.plot
* df.ai.verify
* UDF generation.
output

This PR starts with `df.ai.plot`. After changes, the plotting will be like:
```
[CODE GEN 1]
INFO: Getting the following error: 
[UNRESOLVED_COLUMN.WITH_SUGGESTION] A column or function parameter with name `start_date` cannot be resolved. Did you mean one of the following? [`start_lat`, `start_lon`, `end_lat`, `end_lon`, `airline`, `airport1`, `airport2`, `cnt`].
INFO: Retrying with 3 retries left
INFO: I apologize for the confusion. It seems there was a misunderstanding in the requirements. If the DataFrame doesn't have a 'start_date' column, we can't filter for flights in February 2011. 
[CODE GEN 2]
...
```
Also, this PR:
1. improves the prompt of plotting
2. changes the return type of `df.ai.plot` from None to str, so that it is easier for users to get the generated code, and easier to test plotting.

### CODE REVIEW

File: pyspark_ai/prompt.py [hunk 1]
```diff
@@ -227,31 +227,27 @@
...
-You are an Apache Spark SQL expert programmer.
-It is forbidden to include old deprecated APIs in your code.
 For example, you will not use the pandas method "append" because it is deprecated.
 
 Given a pyspark DataFrame `df`, with the output columns:
 {columns}
 
-And an explanation of `df`: {explain}
+Write Python code to visualize the result of `df` using plotly:
+1. Do any aggregation against `df` first, before converting the `df` to a pandas DataFrame.
```
Comment (gengliangwang) [inline pyspark_ai/prompt.py:236]: I am using bullet points to make instructions more clear.

File: pyspark_ai/prompt.py [hunk 2]
```diff
@@ -227,31 +227,25 @@
...
-Do not use scatter plot to display any kind of percentage data.
-You must import and start your Spark session if you use a Spark DataFrame.
-Remember to ensure that your code does NOT include "append" anywhere, under any circumstance (use pd.concat instead).
+Write Python code to visualize the result of `df` using plotly:
+1. Do any aggregation against `df` first, before converting the `df` to a pandas DataFrame. 
+2. Make sure to use the exact column names of `df`.
+3. Your code may NOT contain "append" anywhere. Instead of append, use pd.concat.
+4. There is no need to install any package with pip. Do include any necessary import statements.
+5. Display the plot directly, instead of saving into an HTML.
+6. Do not use scatter plot to display any kind of percentage data.
```
Comment (xinrong-meng) [inline pyspark_ai/prompt.py:239]: May I ask what's the reason for 6.?

Comment (gengliangwang) [reply pyspark_ai/prompt.py:239]: This PR is for refactoring. cc @asl3 @allisonwang-db do you recall the reason for this one?

File: tests/data/2011_february_aa_flight_paths.csv [hunk 3]
```diff
@@ -0,0 +1,179 @@
+start_lat,start_lon,end_lat,end_lon,airline,airport1,airport2,cnt
```
Comment (xinrong-meng) [inline tests/data/2011_february_aa_flight_paths.csv:1]: I'm wondering why we don't pull the data from the original website but copy the file.

File: pyspark_ai/python_executor.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        response = self._generate_python_with_retries(
+            self.df, self.llm, messages, self.max_retries
+        )
+        if use_cache and self.cache is not None:
+            self.cache.update(prompt_str, response)
+        return response
+
+    @staticmethod
+    def _execute_code(df: DataFrame, code: str):
+        import pandas as pd  # noqa: F401
```
Comment (xinrong-meng) [inline pyspark_ai/python_executor.py:56]: I'm wondering what's that used for

File: pyspark_ai/python_executor.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            return code
+        except Exception as e:
+            if self.logger is not None:
+                self.logger.warning("Getting the following error: \n" + str(e))
+            if retries <= 0:
+                # if we have no more retries, raise the exception
+                self.logger.info(
+                    "No more retries left, please modify the instruction or modify the generated code"
+                )
+                return ""
```
Comment (xinrong-meng) [inline pyspark_ai/python_executor.py:82]: Do we want to raise an exception here instead?

**Comment by xinrong-meng** [comment]
> "changes the return type of df.ai.plot from None to str, so that it is easier for users to get the generated code, and easier to test plotting."

I'm wondering if we should make it configurable to enhance the experience for our current users. If users are only interested in the plots, having generated code could be an unnecessary complication for them.

File: tests/data/2011_february_aa_flight_paths.csv (hunk 3 above)
Comment (gengliangwang) [reply tests/data/2011_february_aa_flight_paths.csv:1]: It's simpler. Is there a license concern?

File: pyspark_ai/python_executor.py (hunk 4 above)
Comment (gengliangwang) [reply pyspark_ai/python_executor.py:56]: Otherwise there can be error of pd not defined

File: pyspark_ai/python_executor.py (hunk 5 above)
Comment (gengliangwang) [reply pyspark_ai/python_executor.py:82]: I thought about it. The error stack trace will make the logs hard to read, especially with retries.

**Comment by gengliangwang** [comment]
> I'm wondering if we should make it configurable to enhance the experience for our current users. If users are only interested in the plots, having generated code could be an unnecessary complication for them.

Note: the plotting will be executed anyway. We are returning the string so that users/tests can get the precise result code.

File: pyspark_ai/python_executor.py (hunk 4 above)
Comment (xinrong-meng) [reply pyspark_ai/python_executor.py:56]: It also failed to import pyspark.sql.functions at times. A prompt improvement might be better than hardcoding imports here.

**Comment by xinrong-meng** [comment]
I understand. I was wondering if making the generated code optional will enhance the user experience.

**Review by xinrong-meng — APPROVED** [review]

**Pull request merged**